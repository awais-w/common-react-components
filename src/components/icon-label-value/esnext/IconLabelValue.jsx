import React from 'react';
import PropTypes, { oneOf } from 'prop-types';
import Icon from '@cmc/icon/esnext/Icon';
import bem from '@argos/utils/esnext/bem';
import isEmpty from '@cmc/utils/esnext/isEmpty';
import ReactTooltip from 'react-tooltip';
import uuid from 'uuid';
import './IconLabelValue.scss';

const cn = bem({ prefix: 'cmc-', block: 'icon-label-value' });

const IconLabelValue = ({ value, icon, isCustom, isInline, label, onClick, tooltipValue, color, labelColor }) => {
  const handleOnClick = (e, index) => {
    // Potentially an onClick on each value item instead of the top level.
    // This if statement helps to handle that, ensuring the correct one is clicked.
    if (!isEmpty(index) && value && value[index] && value[index].onClick) {
      value[index].onClick(e);
    } else if (onClick) {
      onClick(e);
    }
  };

  const valueContent = !isEmpty(value) ? (
    !Array.isArray(value) ? (
      <span className={cn('value', labelColor)}>{value}</span>
    ) : (
      value.map((valueObj, index) => (
        <span
          key={valueObj.key}
          className={cn('value-multi', valueObj.onClick && 'clickable', labelColor)}
          onClick={(e) => handleOnClick(e, index)}>
          {valueObj.value}
        </span>
      ))
    )
  ) : null;

  const tooltipData = tooltipValue
    ? {
        'data-for': uuid.v4(),
        'data-tip': tooltipValue,
      }
    : {};

  return (
    <div
      className={cn(null, { inline: isInline, 'no-icon': isEmpty(icon) })}
      {...tooltipData}
      onClick={(e) => handleOnClick(e)}>
      {!isEmpty(icon) && <Icon isCustom={isCustom} icon={icon} color={color} />}
      {!isEmpty(label) && <span className={cn('label', labelColor)}>{label}</span>}
      {valueContent && valueContent}
      {tooltipValue && <ReactTooltip id={tooltipData['data-for']} effect='solid' />}
    </div>
  );
};

IconLabelValue.propTypes = {
  icon: PropTypes.string,
  label: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.string.isRequired,
        key: PropTypes.string.isRequired,
        onClick: PropTypes.func,
      }),
    ),
  ]),
  onClick: PropTypes.func,
  isInline: PropTypes.bool,
  isCustom: PropTypes.bool,
  tooltipValue: PropTypes.string,
  labelColor: oneOf(['amber', 'blue', 'red', 'green', 'grey']),
};
export default IconLabelValue;
