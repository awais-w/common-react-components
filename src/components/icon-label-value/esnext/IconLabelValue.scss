@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-icon-label-value {
  @include font-standard;
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  column-gap: $half;

  .material-icons {
    padding-right: $half;
    user-select: none;
  }

  &--inline {
    margin-right: $base;
    display: inline-flex;

    // double-click selection fix for inline-blocks
    &::after {
      content: '';
      display: block;
    }
  }

  &--no-icon {
    &::before {
      content: '';
      height: $large;
      margin-left: -$half;
    }
  }

  &__label {
    @include font-bold;
  }

  &__label,
  &__value {
    line-height: initial;
  }

  &__value-multi {
    &--clickable {
      cursor: pointer;
      color: $color-blue-cobalt;
      text-decoration: underline;

      &:hover {
        color: $primary-color-light;
      }
    }
  }

  &__label,
  &__value {
    &--amber,
    &.amber {
      color: $color-sainsburys;
    }
    &--blue,
    &.blue {
      color: $primary-color;
    }
    &--red,
    &.red {
      color: $color-rust;
    }
    &--green,
    &.green {
      color: $color-green;
    }
    &--grey,
    &.grey {
      color: $color-grey-dark;
    }
  }
}
