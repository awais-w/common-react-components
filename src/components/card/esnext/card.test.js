import React from 'react';
import { shallow } from 'enzyme';
import IconLabelValue from '@cmc/icon-label-value/esnext';
import { stub } from 'sinon';
import Skeleton from 'react-loading-skeleton';
import Card from './card';

const defaultProps = {
  title1: 'Title 1',
  title2: 'Title 2',
  status: 'DELIVERED',
  isExpanded: false,
};

const setup = ({ props }) => {
  const component = shallow(<Card {...props} />);

  return {
    wrapper: component,
    getCard: () => component.find('.ac-panel'),
    getHeading: () => component.find('.ac-panel__header'),
    getTitleWrapper: () => component.find('.ac-panel__title'),
    getTitle1: () => component.find('.ac-panel__title1'),
    getTitle2: () => component.find('.ac-panel__title2'),
    getTitle3: () => component.find('.ac-panel__title3'),
    getExpandClass: () => component.find('.expandable-panel'),
    getClickClass: () => component.find('.clickable-panel'),
    getIcon: () => component.find('.material-icons'),
    getRightIcons: () => component.find('ac-panel__header-right').find(IconLabelValue),
    getSummary: () => component.find('.ac-panel__summary'),
    getBody: () => component.find('.ac-panel__body'),
    getChildren: () => component.find('.child'),
    getActionsWrapper: () => component.find('.ac-panel__actions'),
    getActionsIcon: () => component.find('.ac-panel__actions-icon'),
    getActionsMenu: () => component.find('.ac-panel__actions-menu'),
    getMenuState: () => component.state('isMenuOpen'),
    getSkeleton: () => component.find(Skeleton),
  };
};

describe('<Card />', () => {
  describe('with default props', () => {
    const component = setup({ props: defaultProps });

    it('should match default snapshot', () => {
      expect(component.wrapper).toMatchSnapshot();
    });

    it('should render a <Card /> wrapper', () => {
      expect(component.getCard().exists()).toBe(true);
    });

    it('should render a heading', () => {
      expect(component.getHeading().exists()).toBe(true);
    });

    it('should render a <H2 /> title wrapper', () => {
      expect(component.getTitleWrapper().exists()).toBe(true);
      expect(component.getTitleWrapper().type()).toEqual('h3');
    });

    it('should render title1', () => {
      expect(component.getTitle1().exists()).toBe(true);
    });

    it('should render title2', () => {
      expect(component.getTitle2().exists()).toBe(true);
    });

    it('should not be expandable', () => {
      expect(component.getExpandClass().exists()).toBe(false);
    });

    it('should not be clickable', () => {
      expect(component.getClickClass().exists()).toBe(false);
    });
  });

  describe('with custom props', () => {
    describe('with a title3 prop', () => {
      const title3 = 'wumbo';
      const component = setup({ props: { ...defaultProps, title3 } });

      it('should render a title3', () => {
        expect(component.getTitle3().exists()).toBe(true);
      });
    });

    describe('with icon prop', () => {
      const icon = 'wumbo';
      const component = setup({ props: { ...defaultProps, icon } });

      it('should render an icon', () => {
        expect(component.getIcon().exists()).toBe(true);
      });
    });

    describe('with icons on the right side', () => {
      const rightIcons = [
        { icon: 'gabriel', label: 'icon1 label', tooltip: 'Hello', className: 'classy', onClick: stub },
        { icon: 'awais', label: 'icon2 label', tooltip: 'World', className: 'superClassy', onClick: stub },
      ];
      const component = setup({ props: { rightIcons } });
      it('should render as many icons as passed', () => {
        component.getRightIcons().forEach((iconFound, index) => {
          expect(iconFound.prop('icon')).toBe(rightIcons[index].icon);
          expect(iconFound.prop('label')).toBe(rightIcons[index].label);
          expect(iconFound.prop('tooltipValue')).toBe(rightIcons[index].tooltip);
          expect(iconFound.prop('className')).toBe(rightIcons[index].className);
          expect(iconFound.prop('onClick')).toBe(rightIcons[index].onClick);
        });
      });
    });

    describe('with summary prop', () => {
      const summary = 'I am a summary';

      describe('when <Card /> is not expanded', () => {
        const component = setup({ props: { ...defaultProps, summary } });

        it('should render a summary', () => {
          expect(component.getSummary().exists()).toBe(true);
          expect(component.getSummary().text()).toEqual(summary);
        });
      });

      describe('when <Card /> is expanded', () => {
        const isExpanded = true;
        const component = setup({ props: { ...defaultProps, summary, isExpanded } });

        it('should not render a summary', () => {
          expect(component.getSummary().exists()).toBe(false);
        });
      });
    });

    describe('with children', () => {
      const children = [
        <div key='1' className='child'>
          Child 1
        </div>,
        <div key='2' className='child'>
          Child 2
        </div>,
        <div key='3' className='child'>
          Child 3
        </div>,
      ];

      describe('when <Card /> is not expanded', () => {
        const component = setup({ props: { ...defaultProps, children } });

        it('should not render any children', () => {
          expect(component.getBody().exists()).toBe(false);
        });
      });

      describe('when <Card /> is expanded', () => {
        const isExpanded = true;
        const component = setup({ props: { ...defaultProps, children, isExpanded } });

        it('should render children props', () => {
          expect(component.getBody().exists()).toBe(true);
          expect(component.getChildren().length).toEqual(3);
        });
      });
    });

    describe('with isAsyncBody prop', () => {
      const isAsyncBody = true;
      const component = setup({ props: { ...defaultProps, isAsyncBody } });

      it('should be expandable', () => {
        expect(component.getExpandClass().exists()).toBe(true);
      });
    });

    describe('with actions prop', () => {
      const onToggle = jest.fn();
      const actions = (
        <div>
          <div key='1' className='cmc-panel__actions-item'>
            Action 1
          </div>
          ,
          <div key='2' className='cmc-panel__actions-item'>
            Action 2
          </div>
          ,
          <div key='3' className='cmc-panel__actions-item'>
            Action 3
          </div>
          ,
        </div>
      );

      const component = setup({ props: { ...defaultProps, actions, onToggle } });

      it('should render an actions wrapper', () => {
        expect(component.getActionsWrapper().exists()).toBe(true);
      });

      it('should render an actions icon', () => {
        expect(component.getActionsIcon().exists()).toBe(true);
      });

      it('should render an actions menu', () => {
        expect(component.getActionsMenu().exists()).toBe(true);
      });

      it('should not have an open menu', () => {
        expect(component.getMenuState()).toBe(false);
      });

      describe('when actions menu is clicked', () => {
        it('should have an open menu', () => {
          component.getActionsWrapper().simulate('click', {
            preventDefault: () => {},
            stopPropagation: () => {},
          });
          expect(component.getMenuState()).toBe(true);
        });
      });
    });
  });

  describe('when <Card /> heading is clicked', () => {
    const onToggle = jest.fn();
    const children = [
      <div key='1' className='child'>
        Child 1
      </div>,
      <div key='2' className='child'>
        Child 2
      </div>,
      <div key='3' className='child'>
        Child 3
      </div>,
    ];

    const component = setup({ props: { ...defaultProps, children, onToggle } });

    it('should call the onToggle function prop', () => {
      component.getHeading().simulate('click');
      expect(onToggle).toBeCalled();
    });
  });

  describe('when isLoading', () => {
    const component = setup({ props: defaultProps });
    it('renders a skeleton', () => {
      component.wrapper.setProps({ isLoading: true });
      expect(component.getSkeleton().exists()).toBeTruthy();
    });
  });
});
