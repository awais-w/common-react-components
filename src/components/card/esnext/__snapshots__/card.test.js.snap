// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> with default props should match default snapshot 1`] = `
<section
  className="ac-panel panel panel-collapsed status-delivered"
>
  <div
    className="ac-panel__header panel-heading"
  >
    <div
      className="ac-panel__header-wrapper"
    >
      <span
        className="ac-panel__icon"
      >
        <Icon
          icon="icon-delivered"
          id="icon-delivered"
          isCustom={true}
          onClick={null}
        />
      </span>
      <h3
        className="ac-panel__title"
      >
        <span
          className="ac-panel__title1 title-item"
        >
          Title 1
        </span>
        <span
          className="ac-panel__title2 title-item"
        >
          Title 2
        </span>
      </h3>
    </div>
  </div>
</section>
`;
